golang/cosmos/vm/server.go

```diff
@@ -1,23 +1,93 @@
 package vm
 
 import (
+	"context"
 	"fmt"
+	"sync"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-type AgdServer struct{}
+const VmPortMappingContextKey = sdk.ContextKey("vm-port-mapping")
+
+type AgdServer struct {
+	currentCtx    context.Context
+	mtx           sync.Mutex
+	lastPort      int
+	portToHandler map[int]PortHandler
+	portToName    map[int]string
+	nameToPort    map[string]int
+}
+
+var wrappedEmptySDKContext = sdk.WrapSDKContext(
+	sdk.Context{}.WithContext(context.Background()),
+)
 
 func NewAgdServer() *AgdServer {
-	return &AgdServer{}
+	return &AgdServer{
+		currentCtx:    wrappedEmptySDKContext,
+		mtx:           sync.Mutex{},
+		portToHandler: make(map[int]PortHandler),
+		portToName:    make(map[int]string),
+		nameToPort:    make(map[string]int),
+	}
+}
+
+func (s *AgdServer) SetControllerContext(ctx sdk.Context) func() {
+	// We are only called by the controller, so we assume that it is billing its
+	// own meter usage.
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	s.currentCtx = sdk.WrapSDKContext(ctx.WithGasMeter(sdk.NewInfiniteGasMeter()))
+	return func() {
+		s.mtx.Lock()
+		defer s.mtx.Unlock()
+		s.currentCtx = wrappedEmptySDKContext
+	}
+}
+
+func (s *AgdServer) getContextAndHandler(port int) (context.Context, PortHandler) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	ctx := s.currentCtx
+	handler := s.portToHandler[port]
+	return ctx, handler
 }
 
 // ReceiveMessage is the method the VM calls in order to have agd receive a
 // Message.
-func (s AgdServer) ReceiveMessage(msg *Message, reply *string) error {
-	handler := portToHandler[msg.Port]
+func (s *AgdServer) ReceiveMessage(msg *Message, reply *string) error {
+	ctx, handler := s.getContextAndHandler(msg.Port)
 	if handler == nil {
 		return fmt.Errorf("unregistered port %d", msg.Port)
 	}
-	resp, err := handler.Receive(controllerContext, msg.Data)
+	resp, err := handler.Receive(ctx, msg.Data)
 	*reply = resp
 	return err
 }
+
+func (s *AgdServer) GetPort(name string) int {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.nameToPort[name]
+}
+
+func (s *AgdServer) RegisterPortHandler(name string, portHandler PortHandler) int {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	s.lastPort++
+	s.portToHandler[s.lastPort] = NewProtectedPortHandler(portHandler)
+	s.portToName[s.lastPort] = name
+	s.nameToPort[name] = s.lastPort
+	return s.lastPort
+}
+
+func (s *AgdServer) UnregisterPortHandler(portNum int) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	delete(s.portToHandler, portNum)
+	name := s.portToName[portNum]
+	delete(s.portToName, portNum)
+	delete(s.nameToPort, name)
```

Comment:
Potential problem here if a second port handler is registered for the same name and you unregister the first one's port number - it still removes the name-to-new-port mapping. Consider either:

- making it an error to register a handler for an already-registered name;
- registering an already-registered name merely overwrites the handler and uses the old port number;
- registering an already-registered name completely unregisters the old one before registering the new one.

Reply:
Thanks, I made it an error to register a handler for an already-registered name.